pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -349,7 +349,26 @@ func (c *Cluster) startOSDDaemon(nodeName string, config *provisionConfig, confi
 				continue
 			}
 			logger.Infof("deployment for osd %d already exists", osd.ID)
+			existingDeployment, err := c.context.Clientset.Extensions().Deployments(c.Namespace).Get(dp.Name, metav1.GetOptions{})
```

Comment:
as discussed, I think we should try to update the deployment if it already exists, giving us free automatic upgrade basically.  As long as the `Deployment.Update` call doesn't cause disruption by recycling the OSD pods when nothing has actually changed, which you will be testing, then I think that would be a big win.  The same could be said for deployments for other components such as MGR, MDS, RGW.

Reply:
updating the deployment does the right thing here. If the spec didn't change, the running osd pod continues running and there is no disruption. If the pod spec does change (ie. version update), the pod will be recycled as expected. The only side effect I see from updating when the pod spec doesn't change is that the "ObservedGeneration" of the deployment is incremented, which simply allows us to detect that the new spec is in a good state and the operator can continue.